### ISSUE TITLE
can't distinguish identical local and codebase definitions in TDNR

### ISSUE THREAD

**Issue by aryairani** [unisonweb/unison#377]
Given this program:

```haskell
Sequence.eqWith : (a -> a -> Boolean) -> [a] -> [a] -> Boolean
Sequence.eqWith f s1 s2 = false

(NatSequence.==) : [Nat] -> [Nat] -> Boolean
(NatSequence.==) = eqWith (Nat.==)

-- There's currently no built-in `(Boolean.==)`
(Boolean.==) : Boolean -> Boolean -> Boolean
a Boolean.== b = case (a,b) of
  (true, true) -> true
  (false, false) -> true
  _ -> false
```

where all of the above definitions have already been added to the codebase, the presence of local and repo versions of `Sequence.eqWith` produce an ambiguity that TDNR can't resolve.

![image](https://user-images.githubusercontent.com/538571/54076034-7a7a8480-4274-11e9-8ed8-3969ff9b9de0.png)

Suggestion 1: detect that the two candidate definitions are identical and choose whichever one will cause the fewest bugs to manifest
Suggestion 2: if the two definitions are not identical, then display hash-qualified versions in the error message to help distinguish them.  Idea: the hash `#local` is not valid base58, due to the lower-case Ls.  We could interpret `foo#local` to mean the version of `foo` that is defined in this scratch file.  Alternatively, we could just identify them by `foo` vs `foo#<hashfromcodebase>`.  Regardless of how we uniquely identify the two definitions, I think our error messages need to distinguish local vs codebase definitions in many cases.

**Comment by pchiusano** [comment]
Todo:

1. Look at TDNR to see how to indicate the source of different suggestions (.u vs codebase) (maybe `suggestionReplacement` is good enough.  Can look up `v`'s `Referent` to dedup against other suggestions.  Consider switching suggestionName to HashQualified from current namespace or typelookup or whatever we have.  Consider switching `suggestionReplacement` from `Either v Referent` to `Referent` or maybe `Either Referent Referent`.

2. Prove or disprove that resolution can be confusing/inconsistent depending on lexical vs codebase definition source.

3. If a local and codebase definition both match, report the situation to the user.  Recommend a fully-qualified name to use the local definition, or fully-qualified name + hash to use the codebase definition.

**Comment by aryairani** [comment]
@runarorama Hey is this something you'd have expected to have been resolved in the latest wip?

**Comment by aryairani** [comment]
Seems to be fixed in 511a2587c871877780231cfefd87e2ed8f43dd7f.
